Add configuration of the backoff algorithm for the federation client (#15754)
matrix-org/synapse: docs/usage/configuration/config_documentation.md

* `max_short_retries`: maximum number of retries for the short retry algo. Default to 3 attempts.
* `max_long_retries`: maximum number of retries for the long retry algo. Default to 10 attempts.

> The following options are related to configuring the backoff parameters used fo a specific destination.
> Unlike previous configuration those values applies across all requests,
> and the state of the backoff is stored on DB.

Reviewer: Some minor wording nits.
```suggestion
The following options are related to configuring the backoff parameters used for a specific destination.
Unlike previous configuration options, these values apply across all requests
and the state of the backoff is stored in the database.
```

> The following options are related to configuring the backoff parameters used for a specific destination.
> Unlike previous configuration options, these values apply across all requests
> and the state of the backoff is stored in the database.

Reviewer 2: Probably prefer "retry logic" over "backoff" since we use it above and the options have `retry` in name. I think "backoff" makes sense in the second sentence.

```suggestion
The following options control the retry logic when communicating with a specific homeserver destination.
Unlike the previous configuration options, these values apply across all requests
for a given destination and the state of the backoff is stored in the database.
```

> * `destination_min_retry_interval`: the initial backoff, after the first request fails, in seconds. Default to 10mn.
> * `destination_retry_multiplier`: how much we multiply the backoff by after each subsequent fail. Default to 2.
> * `destination_max_retry_interval`: a cap on the backoff. Default to one day.

Reviewer: ```suggestion
* `destination_min_retry_interval`: the initial backoff, after the first request fails, in seconds. Defaults to 10m.
* `destination_retry_multiplier`: how much we multiply the backoff by after each subsequent fail. Defaults to 2.
* `destination_max_retry_interval`: a cap on the backoff. Defaults to one day.
```

Example configuration:
```yaml

> federation:

Reviewer: We should add the new config options to the example below too.
